File: core/src/test/scala/diplomacy/util/DirectedGraphSpec.scala
package diplomacy.util

import diplomacy.UnitSpec

class DirectedGraphSpec extends UnitSpec {
  "A directed graph" when {
    "the type of nodes is Set" can {
      "merge the nodes." in {
        val g1 = DirectedGraph(
          Set(Set(1), Set(2), Set(3), Set(4)),
          Set(Set(1) -> Set(2), Set(2) -> Set(3), Set(3) -> Set(2), Set(3) -> Set(4))
        )
        val g2 = g1.merge(Set(Set(2), Set(3)))
        g2.nodes should be (Set(Set(1), Set(2, 3), Set(4)))
        g2.edges should be(Set(Set(1) -> Set(2, 3), Set(2, 3) -> Set(4)))
      }
    }
    "it has a cycle" should {
      "return the cycle as a return-value of #cycle (1)." in {
        val g = DirectedGraph(Set(1, 2, 3, 4), Set(1 -> 2, 2 -> 3, 3 -> 2, 3 -> 4))
        g.cycle should be (Option(Seq(2, 3)))
      }
      "return the cycle as a return-value of #cycle (2)." in {
        val g = DirectedGraph(Set(1, 2, 3, 4), Set(2 -> 1, 2 -> 3, 3 -> 2, 4 -> 2))
        g.cycle should be (Option(Seq(2, 3)))
      }
    }
    "there is no cycle" should {
      "return None as a return-value of #cycle." in {
        val g = DirectedGraph(Set(1, 2, 4), Set(1 -> 2, 2 -> 4))
        g.cycle should be(None)
      }
    }
  }
  it can {
    "delete a node." in {
      val g1 = DirectedGraph(Set(1, 2, 3, 4), Set(1 -> 2, 2 -> 3, 3 -> 2, 3 -> 4))
      val g2 = g1 - 2

      g2.nodes should be(Set(1, 3, 4))
      g2.edges should be(Set(3 -> 4))
    }
  }
}

File: test/util/directed-graph-spec.ts
import * as chai from "chai"
import { Edge, DirectedGraph } from "./../../src/util/module"

const should = chai.should()

describe("DirectedGraph", () => {
  const nodes = [new Set(["a"]), new Set(["b"]), new Set(["c"]), new Set(["d"])]
  it("can merge the nodes.", () => {
    const g1 = new DirectedGraph<string>(
      [
        new Edge(nodes[0], nodes[1]), new Edge(nodes[1], nodes[2]),
        new Edge(nodes[2], nodes[1]), new Edge(nodes[2], nodes[3])
      ]
    )
    const g2 = g1.mergeNodes(new Set([nodes[1], nodes[2]]))

    Array.from(g2.nodes).should.have.deep.members(
      [new Set(["a"]), new Set(["b", "c"]), new Set(["d"])]
    )
    Array.from(g2.edges).should.have.deep.members(
      [new Edge(new Set(["a"]), new Set(["b", "c"])), new Edge(new Set(["b", "c"]), new Set(["d"]))]
    )

    const g3 = g2.mergeNodes(g2.nodes)
    Array.from(g3.nodes).should.have.deep.members([new Set(["a", "d", "b", "c"])])
  })
  it("can delete a node.", () => {
    const g1 = new DirectedGraph<string>(
      [
        new Edge(nodes[0], nodes[1]), new Edge(nodes[1], nodes[2]),
        new Edge(nodes[2], nodes[1]), new Edge(nodes[2], nodes[3])
      ]
    )
    const g2 = g1.deleteNode(nodes[1])

    Array.from(g2.nodes).should.have.deep.members([new Set(["a"]), new Set(["c"]), new Set(["d"])])
    Array.from(g2.edges).should.have.deep.members([new Edge(new Set(["c"]), new Set(["d"]))])
  })
  describe("#getCycle", () => {
    it("returns a cycle (1)", () => {
      const g = new DirectedGraph<string>(
        [
          new Edge(nodes[0], nodes[1]), new Edge(nodes[1], nodes[2]),
          new Edge(nodes[2], nodes[1]), new Edge(nodes[2], nodes[3])
        ]
      )
      const cycle = g.getCycle()
      should.not.equal(cycle, null)
      if (cycle) {
        cycle.should.have.deep.members([nodes[1], nodes[2]])
      }
    })
    it("returns a cycle (2)", () => {
      const g = new DirectedGraph<string>(
        [
          new Edge(nodes[0], nodes[1]), new Edge(nodes[1], nodes[2]),
          new Edge(nodes[2], nodes[1]), new Edge(nodes[2], nodes[3])
        ]
      )
      const cycle = g.getCycle()
      should.not.equal(cycle, null)
      if (cycle) {
        cycle.should.have.deep.members([nodes[1], nodes[2]])
      }
    })
    it("returns null if there are no cycles", () => {
      const g = new DirectedGraph<string>(
        [new Edge(nodes[0], nodes[1]), new Edge(nodes[1], nodes[3])]
      )
      const cycle = g.getCycle()
      should.equal(cycle, null)
    })
  })
})
